src/chipper/keyboard2.cljs

```diff
@@ -0,0 +1,233 @@
+;; because position is 0-indexed
+(def frame-count 32)
+(def frame-length 32)
+(def chan-count 4)
+(def attr-count 3)
+;; a hack, because of course
+(def total-attr-positions (* chan-count attr-count))
+
+(def -motions
+  {:down  [ 1  0  0]
+   :up    [-1  0  0]
+   :left  [ 0  0 -1]
+   :right [ 0  0  1]})
+
+(def normal-keymap
+  {;; relative movement
+   ;; keeping :position for clarity. The representation for cursor position may
+   ;; change in the future, thus -motions
+   :KeyJ        [:relative-position (:down  -motions)]
+   :KeyK        [:relative-position (:up    -motions)]
+   :KeyH        [:relative-position (:left  -motions)]
+   :KeyL        [:relative-position (:right -motions)]
+
+   :KeyEnter    [:relative-position (:down  -motions)]
+   :ArrowDown   [:relative-position (:down  -motions)]
+   :ArrowUp     [:relative-position (:up    -motions)]
+   :ArrowLeft   [:relative-position (:left  -motions)]
+   :ArrowRight  [:relative-position (:right -motions)]
+
+   :KeyW        [:relative-position [measure-size     0 0]]
+   :KeyB        [:relative-position [(- measure-size) 0 0]]
+
+   :ShiftBracketRight [:frame  1]
+   :ShiftBracketLeft  [:frame -1]
+
+   ;; absolute movement
+   :KeyG        [:absolute-position [:first :first :first]]
+   :ShiftKeyG   [:absolute-position [:last  :last  :last]]
+   :Digit0      [:absolute-position [:first :first :first]]
+   :ShiftDigit4 [:absolute-position [:last  :last  :last]]
+
+   ;; mode change
+   :KeyI [:mode :edit]
+   :KeyV [:mode :visual]
+
+   ;; other
+   :KeyX        [:macro {:type :simple :value [[:attr nil] (:down-line -motions)]}]
+   ;; so these could be special or just, like... :tempo, and :octave
+   ;; I'm leaving them out for now
+   ;; :Period [:special [:tempo 1]]
+   ;; :ShiftPeriod [:special [:tempo -1]]
+   :Space [:special :play-pause]})
+
+(def edit-keymap
+  ;; hm, this isn't going to work quite like you want it to.
+  ;; this is fine for inputting notes...
+  {:KeyA [:attr :C]    :KeyW [:attr :C#]
+   :KeyS [:attr :D]    :KeyE [:attr :D#]
+   :KeyD [:attr :E]    :KeyR [:attr :F]
+   :KeyF [:attr :F#]   :KeyT [:attr :G]
+   :KeyG [:attr :G#]   :KeyY [:attr :A]
+   :KeyH [:attr :A#]   :KeyU [:attr :B]
+
+   ;; ... but when you get to the other attributes, what happens? They have
+   ;; values in [0x00, 0xff].
+   ;; You gotta either have another sort of translation layer where some of your
+   ;; note names become hex digits, or... something
+   ;; In fact, now that I think about it, this is why volume in the pre-alpha version
+   ;; only goes up to 9.
+   :Digit0 [:attr :Digit0]
+   :Digit1 [:attr :Digit1]
+   :Digit2 [:attr :Digit2]
+   :Digit3 [:attr :Digit3]
+   :Digit4 [:attr :Digit4]
+   :Digit5 [:attr :Digit5]
+   :Digit6 [:attr :Digit6]
+   :Digit7 [:attr :Digit7]
+   :Digit8 [:attr :Digit8]
+   :Digit9 [:attr :Digit9]
+
+   :KeyX      :off
+   :ShiftKeyX :stop
+
+   :Backspace [:macro {:type :simple :value [[:attr nil] (:up-line -motions)]}]
```

Comment:
I don't think there's a need to make `:macro` be a map. That'll just add extra unpacking garbage to the handler.

Reply:
what I mean is that if everything else here already is a vec `[:property params]` then macros might as well be the same: `[:macro [:simple seq]]` or `[:macro [:complex something else]]`

Reply:
Oh, well maybe not. 
```
[:macro
 {:type :complex 
  :fn fn
  :optional args
  :go here}]
```
I'll just leave it as is then.